programs/bpf_loader/src/syscalls/mod.rs

```diff
@@ -1907,6 +1838,75 @@ declare_syscall!(
     }
 );
 
+declare_syscallhash!(
+    // Generic Hashing Syscall
+    SyscallHash,
+    fn inner_call<H: HasherImpl>(
+        invoke_context: &mut InvokeContext,
+        vals_addr: u64,
+        vals_len: u64,
+        result_addr: u64,
+        _arg4: u64,
+        _arg5: u64,
+        memory_mapping: &mut MemoryMapping,
+    ) -> Result<u64, Error> {
+        let compute_budget = invoke_context.get_compute_budget();
+        let hash_base_cost = H::get_base_cost(compute_budget);
+        let hash_byte_cost = H::get_byte_cost(compute_budget);
+        let hash_max_slices = H::get_max_slices(compute_budget);
+        if hash_max_slices < vals_len {
+            ic_msg!(
+                invoke_context,
+                "{} Hashing {} sequences in one syscall is over the limit {}",
+                H::NAME,
+                vals_len,
+                hash_max_slices,
+            );
+            return Err(SyscallError::TooManySlices.into());
+        }
+
+        consume_compute_meter(invoke_context, hash_base_cost)?;
+
+        let hash_result = translate_slice_mut::<u8>(
+            memory_mapping,
+            result_addr,
+            std::mem::size_of::<H::Output>() as u64,
+            invoke_context.get_check_aligned(),
+            invoke_context.get_check_size(),
+        )?;
+        let mut hasher = H::create_hasher();
+        if vals_len > 0 {
+            let vals = translate_slice::<&[u8]>(
+                memory_mapping,
+                vals_addr,
+                vals_len,
+                invoke_context.get_check_aligned(),
+                invoke_context.get_check_size(),
+            )?;
+            for val in vals.iter() {
+                let bytes = translate_slice::<u8>(
+                    memory_mapping,
+                    val.as_ptr() as u64,
+                    val.len() as u64,
+                    invoke_context.get_check_aligned(),
+                    invoke_context.get_check_size(),
+                )?;
+                let cost = compute_budget.mem_op_base_cost.max(
+                    hash_byte_cost.saturating_mul(
+                        (val.len() as u64)
+                            .checked_div(2)
+                            .expect("div by non-zero literal"),
+                    ),
+                );
+                consume_compute_meter(invoke_context, cost)?;
+                hasher.hash(bytes);
+            }
+        }
+        hash_result.copy_from_slice(&hasher.result().as_ref());
```

Comment:
i _think_ clippy's going to complain about the `&` here, which is implicit from `.as_ref()`, but let's see what it says before changing